askama_shared/src/escaping.rs

```diff
@@ -47,60 +47,57 @@ where
     }
 }
 
-fn escapable(b: u8) -> bool {
-    match b {
-        b'<' | b'>' | b'&' | b'"' | b'\'' | b'/' => true,
-        _ => false,
-    }
+macro_rules! escaping_body {
+    ($n:expr, $res:ident, $bytes:ident, $start:ident, $quote:expr) => {{
+        let next = $n;
+        if $start < next {
+            $res.extend(&$bytes[$start..next]);
+        }
+        $res.extend($quote);
+        $start = next + 1;
+    }};
 }
 
+const FLAG: u8 = b'>' - b'"';
 pub fn escape(s: String) -> String {
-    let mut found = Vec::new();
+    let mut found = None;
     for (i, b) in s.as_bytes().iter().enumerate() {
-        if escapable(*b) {
-            found.push(i);
+        if b.wrapping_sub(b'"') <= FLAG {
+            match *b {
+                b'<' | b'>' | b'&' | b'"' | b'\'' | b'/' => {
+                    found = Some(i);
+                    break;
+                }
+                _ => (),
+            };
         }
     }
-    if found.is_empty() {
-        return s;
-    }
 
-    let bytes = s.as_bytes();
-    let max_len = bytes.len() + found.len() * 5;
-    let mut res = Vec::<u8>::with_capacity(max_len);
-    let mut start = 0;
-    for idx in &found {
-        if start < *idx {
-            res.extend(&bytes[start..*idx]);
-        }
-        start = *idx + 1;
-        match bytes[*idx] {
-            b'<' => {
-                res.extend(b"&lt;");
-            }
-            b'>' => {
-                res.extend(b"&gt;");
-            }
-            b'&' => {
-                res.extend(b"&amp;");
-            }
-            b'"' => {
-                res.extend(b"&quot;");
-            }
-            b'\'' => {
-                res.extend(b"&#x27;");
-            }
-            b'/' => {
-                res.extend(b"&#x2f;");
+    match found {
+        None => s,
+        Some(found) => {
+            let bytes = s.as_bytes();
+            // Allocate s + (6 * 10% of s)
+            let mut res = Vec::with_capacity(s.len() + 6  * s.len() / 10);
```

Comment:
Actually hardcoding this 10% seems a little too magic, does that make a big difference?

Reply:
2-3% in big table and 5-7% escaping bench, We can can look for another way to adjust the size of the vector. Any more complex heuristics?

Reply:
Okay, so let's just do `s.len() + s.len() / 10`, and maybe add a comment saying that this is a heuristic with a conservative estimate to save on intermediate allocations.

Reply:
We could do more cases in the bench, especially extreme to find the correct heuristic approach. How about?

Reply:
Let's skip that for now. This PR is already complex enough while we get the algorithm sorted out, and adding more cases to the benchmark will likely just make us micro-optimize for the specific benchmarks we've selected.